File: golang/day01/day01.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

type Dial struct {
	pos   int
	zeros int
}

func newDial() *Dial {
	return &Dial{pos: 50}
}

func (dial *Dial) rotate(dir string, dist int) {
	if dir == "L" {
		dial.pos -= dist
	} else {
		dial.pos += dist
	}

	dial.pos = mod(dial.pos, 100)

	if dial.pos == 0 {
		dial.zeros++
	}
}

func (dial *Dial) rotateClicks(dir string, dist int) {
	for i := 0; i < dist; i++ {
		dial.rotate(dir, 1)
	}
}

func rotateDial(clicks bool) int {
	file, err := os.Open("input.txt")

	if err != nil {
		log.Fatalf("failed to open file: %s", err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	dial := newDial()

	for scanner.Scan() {
		line := scanner.Text()

		dist, err := strconv.Atoi(line[1:])

		if err != nil {
			log.Fatalf("failed to parse integer: %s", err)
		}

		if clicks {
			dial.rotateClicks(line[:1], dist)
		} else {
			dial.rotate(line[:1], dist)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("error reading file: %s", err)
	}

	return dial.zeros
}

func mod(a, b int) int {
	return (a%b + b) % b
}

func main() {
	fmt.Println("zeros:", rotateDial(false))
	fmt.Println("zeros with clicks:", rotateDial(true))
}

File: ruby/day01/day01.rb
#! /usr/bin/env ruby

class Dial
  attr_reader :position, :zeros

  def initialize
    @position = 50
    @zeros = 0
  end

  def rotate(direction, distance)
    if direction == 'L'
      @position -= distance
    else
      @position += distance
    end

    @position %= 100
    @zeros += 1 if @position == 0
  end

  def rotate_clicks(direction, distance)
    distance.times do
      rotate(direction, 1)
    end
  end
end

def rotate_dial(clicks: false)
  dial = Dial.new

  File.open('input.txt').each_line do |line|
    direction = line[0]
    distance = line[1..].to_i

    if clicks
      dial.rotate_clicks(direction, distance)
    else
      dial.rotate(direction, distance)
    end
  end

  dial.zeros
end

puts "zeros: #{rotate_dial}"
puts "zeros with clicks: #{rotate_dial(clicks: true)}"
